src/node_buffer.cc

```diff
@@ -602,6 +602,110 @@ void Compare(const FunctionCallbackInfo<Value> &args) {
 }
 
 
+int32_t IndexOf(char* haystack, size_t h_len, char* needle, size_t n_len) {
+  CHECK_GE(h_len, n_len);
+  int32_t ret = -1;
+  size_t i = 0;
+  // TODO(trevnorris): Implement Boyer-Moore string search algorithm.
+  for (; i < h_len - n_len + 1; i++) {
+    if (haystack[i] == needle[0]) {
+      ret = memcmp(haystack + i, needle, n_len);
+      if (ret == 0)
+        break;
+    }
+  }
+  return ret == 0 ? i : -1;
+}
+
+
+void IndexOfString(const FunctionCallbackInfo<Value>& args) {
+  ASSERT(args[0]->IsObject());
+  ASSERT(args[1]->IsString());
+  ASSERT(args[2]->IsNumber());
+
+  ARGS_THIS(args[0].As<Object>());
+  node::Utf8Value str(args.GetIsolate(), args[1]);
+  int32_t offset = args[2]->Int32Value();
+
+  if (offset < 0) {
+    if (offset + static_cast<int32_t>(obj_length) < 0)
+      offset = 0;
+    else
+      offset = obj_length + offset;
+  }
+
+  if (str.length() == 0 ||
+      obj_length == 0 ||
+      (offset != 0 && str.length() + offset <= str.length()) ||
+      str.length() + offset > obj_length)
+    return args.GetReturnValue().Set(-1);
+
+  int32_t r =
+      IndexOf(obj_data + offset, obj_length - offset, *str, str.length());
+  args.GetReturnValue().Set(r == -1 ? -1 : r + offset);
```

Comment:
I haven't been able to see a way it could. should I throw in a check before returning it?